File: alert-service/src/main/java/com/example/alert/util/chartpattern/PivotPointUtils.java
package com.example.alert.util.chartpattern;

import com.example.alert.model.chartpattern.OhlcData;
import java.util.ArrayList;
import java.util.List;

/**
 * Utility class để tìm pivot points trong dữ liệu OHLC
 * Dựa trên thuật toán pivot_points.py
 */
public class PivotPointUtils {
    
    /**
     * Tìm tất cả pivot points trong dữ liệu OHLC
     * @param ohlcData danh sách dữ liệu OHLC
     * @param leftLookback số candle nhìn về phía trước
     * @param rightLookback số candle nhìn về phía sau
     * @return danh sách OhlcData với pivot points được đánh dấu
     */
    public static List<OhlcData> findAllPivotPoints(List<OhlcData> ohlcData, 
                                                   int leftLookback, 
                                                   int rightLookback) {
        List<OhlcData> result = new ArrayList<>(ohlcData);
        
        for (int i = 0; i < result.size(); i++) {
            result.get(i).setPivot(0); // Khởi tạo không phải pivot
            result.get(i).setPivotPos(0.0);
        }
        
        // Tìm pivot highs và pivot lows
        for (int i = leftLookback; i < result.size() - rightLookback; i++) {
            // Kiểm tra pivot high
            if (isPivotHigh(result, i, leftLookback, rightLookback)) {
                result.get(i).setPivot(2); // 2 = pivot high
                result.get(i).setPivotPos(result.get(i).getHigh());
            }
            // Kiểm tra pivot low
            else if (isPivotLow(result, i, leftLookback, rightLookback)) {
                result.get(i).setPivot(1); // 1 = pivot low
                result.get(i).setPivotPos(result.get(i).getLow());
            }
        }
        
        return result;
    }
    
    /**
     * Kiểm tra xem một điểm có phải là pivot high không
     */
    private static boolean isPivotHigh(List<OhlcData> data, int index, 
                                      int leftLookback, int rightLookback) {
        double currentHigh = data.get(index).getHigh();
        
        // Kiểm tra tất cả các điểm bên trái
        for (int i = index - leftLookback; i < index; i++) {
            if (data.get(i).getHigh() >= currentHigh) {
                return false;
            }
        }
        
        // Kiểm tra tất cả các điểm bên phải
        for (int i = index + 1; i <= index + rightLookback; i++) {
            if (data.get(i).getHigh() >= currentHigh) {
                return false;
            }
        }
        
        return true;
    }
    
    /**
     * Kiểm tra xem một điểm có phải là pivot low không
     */
    private static boolean isPivotLow(List<OhlcData> data, int index, 
                                     int leftLookback, int rightLookback) {
        double currentLow = data.get(index).getLow();
        
        // Kiểm tra tất cả các điểm bên trái
        for (int i = index - leftLookback; i < index; i++) {
            if (data.get(i).getLow() <= currentLow) {
                return false;
            }
        }
        
        // Kiểm tra tất cả các điểm bên phải
        for (int i = index + 1; i <= index + rightLookback; i++) {
            if (data.get(i).getLow() <= currentLow) {
                return false;
            }
        }
        
        return true;
    }
}

File: client-service/react-client/src/services/pivotPointUtils.js
/**
 * Pivot Point Utilities
 * Provides functions to find pivot highs and lows in OHLC data
 */

/**
 * Find pivot highs (local maxima) in the data
 * @param {Array} data - Array of OHLCV data with {high, low, open, close, time}
 * @param {number} leftBars - Number of bars to the left
 * @param {number} rightBars - Number of bars to the right
 * @returns {Array} Array of objects {index, value, isPivotHigh: true}
 */
export const findPivotHighs = (data, leftBars = 2, rightBars = 2) => {
  const pivots = [];
  
  for (let i = leftBars; i < data.length - rightBars; i++) {
    let isPivot = true;
    const currentHigh = data[i].high;
    
    // Check left side - all bars must be lower
    for (let j = i - leftBars; j < i; j++) {
      if (data[j].high >= currentHigh) {
        isPivot = false;
        break;
      }
    }
    
    // Check right side - all bars must be lower
    if (isPivot) {
      for (let j = i + 1; j <= i + rightBars; j++) {
        if (data[j].high >= currentHigh) {
          isPivot = false;
          break;
        }
      }
    }
    
    if (isPivot) {
      pivots.push({ 
        index: i, 
        value: currentHigh,
        isPivotHigh: true,
        isPivotLow: false
      });
    }
  }
  
  return pivots;
};

/**
 * Find pivot lows (local minima) in the data
 * @param {Array} data - Array of OHLCV data with {high, low, open, close, time}
 * @param {number} leftBars - Number of bars to the left
 * @param {number} rightBars - Number of bars to the right
 * @returns {Array} Array of objects {index, value, isPivotLow: true}
 */
export const findPivotLows = (data, leftBars = 2, rightBars = 2) => {
  const pivots = [];
  
  for (let i = leftBars; i < data.length - rightBars; i++) {
    let isPivot = true;
    const currentLow = data[i].low;
    
    // Check left side - all bars must be higher
    for (let j = i - leftBars; j < i; j++) {
      if (data[j].low <= currentLow) {
        isPivot = false;
        break;
      }
    }
    
    // Check right side - all bars must be higher
    if (isPivot) {
      for (let j = i + 1; j <= i + rightBars; j++) {
        if (data[j].low <= currentLow) {
          isPivot = false;
          break;
        }
      }
    }
    
    if (isPivot) {
      pivots.push({ 
        index: i, 
        value: currentLow,
        isPivotHigh: false,
        isPivotLow: true
      });
    }
  }
  
  return pivots;
};

/**
 * Find all pivot points (both highs and lows) in the data
 * Returns data with pivot information attached
 * @param {Array} data - Array of OHLCV data
 * @param {number} leftBars - Number of bars to the left
 * @param {number} rightBars - Number of bars to the right
 * @returns {Array} Array with pivot info: {pivot: 1 for high, -1 for low, 0 for none, pivotPos: price}
 */
export const findAllPivotPoints = (data, leftBars = 2, rightBars = 2) => {
  // Create copy of data with pivot info
  const dataWithPivots = data.map(candle => ({
    ...candle,
    pivot: 0,
    pivotPos: null
  }));
  
  // Find pivot highs
  const pivotHighs = findPivotHighs(data, leftBars, rightBars);
  pivotHighs.forEach(pivot => {
    dataWithPivots[pivot.index].pivot = 1;
    dataWithPivots[pivot.index].pivotPos = pivot.value;
  });
  
  // Find pivot lows
  const pivotLows = findPivotLows(data, leftBars, rightBars);
  pivotLows.forEach(pivot => {
    dataWithPivots[pivot.index].pivot = -1;
    dataWithPivots[pivot.index].pivotPos = pivot.value;
  });
  
  return dataWithPivots;
};

/**
 * Filter overlapping patterns based on distance threshold
 * @param {Array} patterns - Array of pattern objects with candleIndex
 * @param {number} minDistance - Minimum distance between patterns (in bars)
 * @returns {Array} Filtered patterns without overlaps
 */
export const filterOverlappingPatterns = (patterns, minDistance = 10) => {
  if (!patterns || patterns.length === 0) return [];
  
  // Sort by candleIndex
  const sorted = [...patterns].sort((a, b) => a.candleIndex - b.candleIndex);
  
  const filtered = [sorted[0]];
  
  for (let i = 1; i < sorted.length; i++) {
    const lastPattern = filtered[filtered.length - 1];
    const currentPattern = sorted[i];
    
    // Check distance
    if (currentPattern.candleIndex - lastPattern.candleIndex >= minDistance) {
      filtered.push(currentPattern);
    }
  }
  
  return filtered;
};
